File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_#{params[:chat_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    Message.create!(
      user_id: data['user_id'],
      chat_id: params[:chat_id],
      content: data['message']
    )
  end
end

File: app/channels/chat_channel.js
import consumer from "./consumer"

document.addEventListener("DOMContentLoaded", () => {
    const chatId = document.getElementById("chat-id").value;
    const chatChannel = consumer.subscriptions.create(
        { channel: "ChatChannel", chat_id: chatId },
        {
            received(data) {
                // Append new message to chat window
                const messagesContainer = document.getElementById("messages");
                messagesContainer.insertAdjacentHTML("beforeend", `<p>${data.content}</p>`);
            },

            speak(message) {
                this.perform("speak", { message: message });
            }
        }
    );

    document.getElementById("message-form").addEventListener("submit", (event) => {
        event.preventDefault();
        const messageInput = document.getElementById("message-input");
        chatChannel.speak(messageInput.value);
        messageInput.value = '';
    });
});
